node-graph/graph-craft/src/document.rs

```diff
@@ -925,6 +925,48 @@ impl NodeNetwork {
 		self.nodes.extend(extraction_nodes);
 	}
 
+	/// Due to the adaptive resolution system, nodes that take a `GraphicGroup` as input must call the upstream node with the `Footprint` parameter.
+	///
+	/// However, in the case of the default input, we must insert a node that takes an input of `Footprint` and returns `GraphicGroup::Empty`, in order to satisfy the type system.
+	/// This is because the standard value node takes in `()`.
+	pub fn resolve_empty_stacks(&mut self) {
+		let new_id = generate_uuid();
+		let mut used = false;
+
+		for node in self.nodes.values_mut() {
+			// Do not do this to the newly inserted empty stack (in case `resolve_empty_stacks` runs multiple times).
+			if node.name == "Empty Stack" {
+				continue;
+			}
+			for input in &mut node.inputs {
+				if !matches!(
+					input,
+					NodeInput::Value {
+						tagged_value: TaggedValue::GraphicGroup(GraphicGroup::EMPTY),
+						..
+					}
+				) {
+					continue;
+				}
+
+				*input = NodeInput::node(new_id, 0);
+				used = true;
```

Comment:
Stylistic suggestion: invert this `if` statement by removing the preceding `!` and put the two lines inside the block in place of the `continue;` to save on a few lines.